### PR TITLE
feat(pvc): support modify ebs

### PR DESCRIPTION
**Pull request by liubog2008**
### What problem does this PR solve?

* Sync volume status when syncing tc status
* Support to config env for the operator when install by helm
* Add a feature gate "VolumeModifying" to enable the volume modifying, default to disable.

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [ ] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
support modify ebs
```

### CODE REVIEW

File: pkg/manager/volumes/delegation/aws/ebs_modifier.go [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+func (m *EBSModifier) getCurrentVolumeStatus(ctx context.Context, id string) (*Volume, error) {
+	res, err := m.c.DescribeVolumesModifications(ctx, &ec2.DescribeVolumesModificationsInput{
+		VolumeIds: []string{id},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, s := range res.VolumesModifications {
+		if s.VolumeId == nil || *s.VolumeId == id {
```
Comment (csuzhangxc) [inline pkg/manager/volumes/delegation/aws/ebs_modifier.go:138]: `*s.VolumeId != id`

File: pkg/manager/volumes/phase.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	size := desired.Size
+	scName := ""
+	if desired.StorageClass != nil {
+		scName = desired.StorageClass.Name
+	}
+
+	return isPVCStatusMatched(pvc, scName, size)
+}
+
+func isPVCStatusMatched(pvc *corev1.PersistentVolumeClaim, scName, size string) bool {
```
Comment (csuzhangxc) [inline pkg/manager/volumes/phase.go:101]: any better name? the returned value is changed not matched. the same for `isPVCSpecMatched` in _pod_vol_modifier.go_

File: pkg/manager/volumes/pod_vol_modifier.go [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+	return &podVolModifier{
+		deps: deps,
+		modifiers: map[string]delegation.VolumeModifier{
+			"ebs.csi.aws.com": aws.NewEBSModifier(deps.AWSConfig),
+		},
+	}
+}
+
+func (p *podVolModifier) Modify(tc *v1alpha1.TidbCluster, pod *corev1.Pod, expected []DesiredVolume, shouldEvictLeader bool) (bool, error) {
+	ctx := context.TODO()
```
Comment (csuzhangxc) [inline pkg/manager/volumes/pod_vol_modifier.go:59]: pass `ctx` in as a parameter?

File: pkg/manager/volumes/pod_vol_modifier.go [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+
+	vol.PVC = updated
+
+	return nil
+}
+
+func (p *podVolModifier) modifyVolume(ctx context.Context, vol *ActualVolume) (bool, error) {
+	m := p.getVolumeModifier(vol.Desired.StorageClass)
+	if m == nil {
+		// skip modifying volume by delegation.VolumeModifier
```
Comment (csuzhangxc) [inline pkg/manager/volumes/pod_vol_modifier.go:462]: log here?

File: pkg/controller/dependences.go [hunk 5]
```diff
@@ -314,6 +320,11 @@ func newDependencies(
 		ingv1beta1Lister = kubeInformerFactory.Extensions().V1beta1().Ingresses().Lister()
 	}
 
+	cfg, err := config.LoadDefaultConfig(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("can't load aws config: %w", err)
+	}
+
```
Comment (KanShiori) [inline pkg/controller/dependences.go:327]: may not return err to be compatible with OP environment.

File: pkg/manager/member/tikv_upgrader.go [hunk 6]
```diff
@@ -178,60 +183,91 @@ func (u *tikvUpgrader) upgradeTiKVPod(tc *v1alpha1.TidbCluster, ordinal int32, n
...
-	_, evicting := upgradePod.Annotations[EvictLeaderBeginTime]
-	if !evicting {
-		return u.beginEvictLeader(tc, storeID, upgradePod)
+	if !done {
+		return controller.RequeueErrorf("upgradeTiKVPod: evicting leader of pod %s for tc %s/%s", upgradePodName, ns, tcName)
 	}
 
-	if u.readyToUpgrade(upgradePod, tc) {
-		mngerutils.SetUpgradePartition(newSet, ordinal)
-		return nil
```
Comment (KanShiori) [inline pkg/manager/member/tikv_upgrader.go:199]: have moved into the func `evictLeaderBeforeUpgrade`

File: pkg/controller/dependences.go (hunk 5 above)
Comment (liubog2008) [reply pkg/controller/dependences.go:327]: check feature gates before loading aws config?

File: pkg/manager/volumes/pod_vol_modifier.go (hunk 4 above)
Comment (liubog2008) [reply pkg/manager/volumes/pod_vol_modifier.go:510]: It's not an invalid path. Size of some volumes can be modified even though delegation.VolumeModifier is not implemented.

File: pkg/manager/volumes/phase.go [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+	// 1. isPVCRevisionChanged: true
+	// 2. needModify: true/false
+	// 3. waitForNextTime: true/false
+	VolumePhaseModifying
+	// 1. isPVCRevisionChanged: false
+	// 2. needModify: false
+	// 3. waitForNextTime: true/false
+	VolumePhaseModified
+
+	VolumePhaseCannotModify
```
Comment (KanShiori) [inline pkg/manager/volumes/phase.go:47]: How about adding some comments to describe every phase?

File: pkg/manager/volumes/pvc_modifier.go [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+const (
+	annoKeyPVCSpecRevision     = "spec.tidb.pingcap.com/revision"
+	annoKeyPVCSpecStorageClass = "spec.tidb.pingcap.com/storage-class"
+	annoKeyPVCSpecStorageSize  = "spec.tidb.pingcap.com/storage-size"
+
+	annoKeyPVCStatusRevision     = "status.tidb.pingcap.com/revision"
+	annoKeyPVCStatusStorageClass = "status.tidb.pingcap.com/storage-class"
+	annoKeyPVCStatusStorageSize  = "status.tidb.pingcap.com/storage-size"
+
+	annoKeyPVCLastTransitionTimestamp = "status.tidb.pingcap.com/last-transition-timestamp"
```
Comment (KanShiori) [inline pkg/manager/volumes/pvc_modifier.go:45]: how about adding some comments to describe every annotation?

File: pkg/manager/volumes/pvc_modifier.go [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+		return nil, fmt.Errorf("failed to list Pods: %w", err)
+	}
+
+	sort.Slice(pods, func(i, k int) bool {
+		a, b := pods[i].Name, pods[k].Name
+		if len(a) != len(b) {
+			return len(a) < len(b)
+		}
+		return a < b
+	})
```
Comment (KanShiori) [inline pkg/manager/volumes/pvc_modifier.go:168]: how about adding comments to describe the reason for the sorting?

**Review by csuzhangxc — APPROVED** [review]

**Review by KanShiori — APPROVED** [review]

File: docs/api-references/docs.md [hunk 10]
```diff
@@ -9168,20 +9168,70 @@ k8s.io/apimachinery/pkg/api/resource.Quantity
...
+<tr>
+<td>
+<code>modifiedStorageClass</code></br>
+<em>
+string
+</em>
+</td>
+<td>
+<em>(Optional)</em>
+<p>ModifiedStorageClass is the modified storage calss of the volume.</p>
```
Comment (coderplay) [inline docs/api-references/docs.md:9210]: incorrect description?

File: docs/api-references/docs.md [hunk 11]
```diff
@@ -9168,20 +9168,70 @@ k8s.io/apimachinery/pkg/api/resource.Quantity
...
 <td>
+<em>(Optional)</em>
 <p>CurrentCapacity is the current capacity of the volume.
 If any volume is resizing, it is the capacity before resizing.
 If all volumes are resized, it is the resized capacity and same as desired capacity.</p>
 </td>
 </tr>
 <tr>
 <td>
+<code>modifiedCapacity</code></br>
```
Comment (coderplay) [inline docs/api-references/docs.md:9179]: The terms "resizedCapacity" and "modifiedCapacity" are confusingly named.

**Pull request merged**